File: data/src/main/java/org/cryptomator/data/cloud/crypto/DirIdCacheFormatPre7.java
package org.cryptomator.data.cloud.crypto;

import android.util.LruCache;

import org.cryptomator.domain.CloudFile;

import java.util.Date;

class DirIdCacheFormatPre7 implements DirIdCache {

	private static final int MAX_SIZE = 1024;

	private final LruCache<DirIdCacheKey, DirIdInfo> cache = new LruCache<>(MAX_SIZE);

	DirIdCacheFormatPre7() {
	}

	public DirIdInfo get(CryptoFolder folder) {
		return cache.get(DirIdCacheKey.toKey(folder));
	}

	public DirIdInfo put(CryptoFolder folder, DirIdInfo dirIdInfo) {
		DirIdCacheKey key = DirIdCacheKey.toKey(folder);
		cache.put(key, dirIdInfo);
		cache.remove(key.withoutModified());
		return dirIdInfo;
	}

	public void evict(CryptoFolder folder) {
		DirIdCacheKey key = DirIdCacheKey.toKey(folder);
		cache.remove(key);
		cache.remove(key.withoutModified());
	}

	@Override
	public void evictSubFoldersOf(CryptoFolder cryptoFolder) {
		// no implementation needed
	}

	private static class DirIdCacheKey {

		static DirIdCacheKey toKey(CryptoFolder folder) {
			return new DirIdCacheKey(folder.getDirFile());
		}

		private final String path;
		private final Date modified;

		private DirIdCacheKey(CloudFile dirFile) {
			this.path = dirFile == null ? null : dirFile.getPath();
			this.modified = dirFile == null ? null : dirFile.getModified().orElse(null);
		}

		private DirIdCacheKey(String path) {
			this.path = path;
			this.modified = null;
		}

		DirIdCacheKey withoutModified() {
			return new DirIdCacheKey(path);
		}

		@Override
		public boolean equals(Object obj) {
			if (obj == this)
				return true;
			if (obj == null || getClass() != obj.getClass())
				return false;
			return internalEquals((DirIdCacheKey) obj);
		}

		private boolean internalEquals(DirIdCacheKey o) {
			return (path == null ? o.path == null : path.equals(o.path)) //
					&& (modified == null ? o.modified == null : modified.equals(o.modified));
		}

		@Override
		public int hashCode() {
			final int prime = 31;
			int hash = 1940604225;
			hash = hash * prime + (path == null ? 0 : path.hashCode());
			hash = hash * prime + (modified == null ? 0 : modified.hashCode());
			return hash;
		}
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/crypto/DirIdCacheFormatPre7.kt
package org.cryptomator.data.cloud.crypto

import android.util.LruCache
import org.cryptomator.data.cloud.crypto.DirIdCache.DirIdInfo
import org.cryptomator.domain.CloudFile
import java.util.Date

internal class DirIdCacheFormatPre7 : DirIdCache {

	private val cache = LruCache<DirIdCacheKey, DirIdInfo>(MAX_SIZE)

	override fun get(folder: CryptoFolder): DirIdInfo? {
		return cache[DirIdCacheKey.toKey(folder)]
	}

	override fun put(folder: CryptoFolder, dirIdInfo: DirIdInfo): DirIdInfo {
		val key = DirIdCacheKey.toKey(folder)
		cache.put(key, dirIdInfo)
		cache.remove(key.withoutModified())
		return dirIdInfo
	}

	override fun evict(folder: CryptoFolder) {
		val key = DirIdCacheKey.toKey(folder)
		cache.remove(key)
		cache.remove(key.withoutModified())
	}

	override fun evictSubFoldersOf(cryptoFolder: CryptoFolder) {
		// no implementation needed
	}

	private class DirIdCacheKey {

		private val path: String?
		private val modified: Date?

		private constructor(dirFile: CloudFile?) {
			path = dirFile?.path
			modified = dirFile?.modified
		}

		private constructor(path: String?) {
			this.path = path
			modified = null
		}

		fun withoutModified(): DirIdCacheKey {
			return DirIdCacheKey(path)
		}

		override fun equals(other: Any?): Boolean {
			if (other === this) {
				return true
			}
			return if (other == null || javaClass != other.javaClass) {
				false
			} else internalEquals(other as DirIdCacheKey)
		}

		private fun internalEquals(o: DirIdCacheKey): Boolean {
			return ((if (path == null) o.path == null else path == o.path) //
					&& if (modified == null) o.modified == null else modified == o.modified)
		}

		override fun hashCode(): Int {
			val prime = 31
			var hash = 1940604225
			hash = hash * prime + (path?.hashCode() ?: 0)
			hash = hash * prime + (modified?.hashCode() ?: 0)
			return hash
		}

		companion object {

			fun toKey(folder: CryptoFolder): DirIdCacheKey {
				return DirIdCacheKey(folder.dirFile)
			}
		}
	}

	companion object {

		private const val MAX_SIZE = 1024
	}
}
